Orange/widgets/unsupervised/owlouvainclustering.py

```diff
@@ -0,0 +1,361 @@
+from collections import deque
+from concurrent.futures import Future  # pylint: disable=unused-import
+from enum import Enum
+
+import networkx as nx  # pylint: disable=unused-import
+import numpy as np
+from AnyQt.QtCore import Qt, pyqtSignal as Signal, QObject
+from AnyQt.QtWidgets import QSlider, QCheckBox, QWidget  # pylint: disable=unused-import
+from types import SimpleNamespace as namespace
+from typing import Optional  # pylint: disable=unused-import
+
+from Orange.clustering.louvain import table_to_knn_graph, Louvain
+from Orange.data import Table, DiscreteVariable
+from Orange.projection import PCA
+from Orange.widgets import widget, gui
+from Orange.widgets.settings import DomainContextHandler, ContextSetting, \
+    Setting
+from Orange.widgets.utils.annotated_data import get_next_name, add_columns, \
+    ANNOTATED_DATA_SIGNAL_NAME
+from Orange.widgets.utils.concurrent import ThreadExecutor
+from Orange.widgets.utils.signals import Input, Output
+from Orange.widgets.widget import Msg
+
+try:
+    from orangecontrib.network.network import Graph
+except:
+    Graph = None
+
+
+_MAX_PCA_COMPONENTS = 50
+_DEFAULT_PCA_COMPONENTS = 25
+_MAX_K_NEIGBOURS = 200
+_DEFAULT_K_NEIGHBORS = 30
+
+
+METRICS = [('Euclidean', 'l2'), ('Manhattan', 'l1')]
```

Comment:
Are these the only two metrics that are supported?

Reply:
For the time being, only these two metrics are supported because of the way sklearn computes nearest neighbours. The only other metrics I had considered adding was the cosine, pearsonr and spearmanr distances, but unfortunately the fast tree methods for finding nearest neighbours in sklearn don't support these and fall back to the O(n^2) brute force approach, which is prohibitively expensive for larger data sets, so we decided it is better to leave them out.

If you feel any of the other supported metrics should be added, they are listed in the source: [KDTree](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/kd_tree.pyx#L14-L15) and [BallTree](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/neighbors/ball_tree.pyx#L13-L22). Most of these don't appear anywhere else in Orange, so including them doesn't really make sense.